src/oc_erchef/apps/chef_index/src/chef_batch_worker.erl

```diff
@@ -0,0 +1,207 @@
+          job_fun :: fun()
+         }).
+
+%%===================================================================
+%% API
+%%===================================================================
+%%
+%% Example Usage
+%%
+%% Fun = fun(SomeArgs) -> do_stuff(SomeArgs) end,
+%% {ok, Pid} = chef_batch_worker:start(Fun, Config)
+%% chef_batch_worker:add(Pid, JobId0, JobArgs0)
+%% chef_batch_worker:add(Pid, JobId1, JobArgs1)
+%% chef_batch_worker:add(Pid, JobId2, JobArgs2)
+%% Result = chef_batch_worker:wait()
+%%
+
+-spec start_link(fun(), proplists:proplist()) -> {ok, pid()} | ignored | {error, term()}.
+start_link(Fun, Config) ->
+    %% Use start_link/3 here as the caller
+    gen_server:start_link(?MODULE, [Fun, Config], []).
+
+-spec add(pid(), term(), list()|term()) -> ok | {error, wait_called}.
+add(Pid, JobName, JobArgs) when is_list(JobArgs)->
+    gen_server:call(Pid, {add_job, JobName, JobArgs});
+add(Pid, JobName, JobArgs) ->
+    add(Pid, JobName, [JobArgs]).
+
+-spec wait(pid()) -> {ok, list()}|{error, already_waiting}|{error, list(), list()}.
+wait(Pid) ->
+    gen_server:call(Pid, gather, infinity).
+
+-spec state(pid()) -> #chef_batch_worker_state{}.
+state(Pid) ->
+    gen_server:call(Pid, get_state).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+init([Fun, _Config]) ->
+    process_flag(trap_exit, true),
+    {ok, #chef_batch_worker_state{job_fun = Fun}}.
+
+handle_call({add_job, JobName, JobArgs}, _From, State = #chef_batch_worker_state{waiting_pid = undefined}) ->
+    {ok, State1} = spawn_worker(JobName, JobArgs, State),
+    {reply, ok, State1};
+%% We don't allow jobs to be added once someone is waiting
+handle_call({add_job, _Data}, _From, State = #chef_batch_worker_state{waiting_pid = _Pid}) ->
+    {reply, {error, wait_called}, State};
+%% If someone calls wait and we are already done, return immediately
+handle_call(gather, _From, #chef_batch_worker_state{waiting_pid = undefined,
+                                                    waiting_workers = [],
+                                                    done_jobs = Done,
+                                                    failed_jobs = Failed}) ->
+    {stop, normal, make_gather_reply(Done, Failed), #chef_batch_worker_state{}};
+%% If someone calls wait and we aren't done, store their pid so we can reply later
+handle_call(gather, From, State = #chef_batch_worker_state{waiting_pid = undefined,
+                                                           waiting_workers = _StillWaiting}) ->
+    {noreply, State#chef_batch_worker_state{waiting_pid = From}};
+%% If someone is already waiting, error
+handle_call(gather, _From, State = #chef_batch_worker_state{waiting_pid = _Pid}) ->
+    {reply, {error, already_waiting}, State};
+%% If a job is done, remove it from the waiting list.  If this is the
+%% last job, rely to the waiting pid if we have one.
```

Comment:
Got the hunch that this would read nice if "last job" was part of the pattern match in the function head. (But I believe you've probably explored that route already.)